TLS related enhancements

lebr0nli opened the pull request:
This PR bring some bugs fix and enhancements to some TLS/multi-threaded related features:
Including:
- Use `pthread_self()` instead of the trick with `__errno_location()` when we can't find TLS base with register, since the trick with `__errno_location()` is unreliable.
- Consolidate heap heuristics checks about TLS within the `pwndbg.gdblib.tls` module for better organization
- Bug fix for the `errno` command (#1535)
- Enhance the strategy for the heap heuristic of `tcache` (#1530)
- Add tests for heap heuristics with multi-threaded cases

lebr0nli committed:
Refactor TLS module

- Replace unreliable `__errno_location()` trick with `pthread_self()` to acquire TLS address
- Consolidate heap heuristics checks about TLS within the `pwndbg.gdblib.tls` module for better organization

lebr0nli committed:
Bug fix for the `errno` command

Calling `__errno_location()` without locking the scheduler can cause another thread to inadvertently continue execution

lebr0nli committed:
Refactor code about heap heuristics of thread-local variables

- Replace some checks with some functions in `pwndbg.gdblib.tls`
- Try to find tcache with `mp_.sbrk_base + 0x10` if the target is single-threaded

lebr0nli committed:
Add tests for heap heuristics with multi-threaded

gsingh93 commented:
> Bug fix for the errno command

This is a neat bug, thanks for catching and fixing it. To prevent this from happening in the future, I'm wondering if we should always be calling `set scheduler-locking on` in `ArgparsedCommand` before a command is executed and `set scheduler-locking off` when the command is finished. Do you think doing this would cause any problems? If so, the way you've done it is fine, we'll just have to be more mindful in future reviews to catch things like this.

disconnect3d commented on pwndbg/commands/misc.py:
```diff
@@ -44,7 +45,8 @@ def errno_(err) -> None:
                 ):
-                    err = int(gdb.parse_and_eval("*((int *(*) (void)) __errno_location) ()"))
+                    with lock_scheduler():
+                        err = int(gdb.parse_and_eval("*((int *(*) (void)) __errno_location) ()"))
```
Do I understand this correctly that if you call a function like this on behalf of the target in GDB, then if you have other threads those other threads are also resumed for the time this function is executed?

Wow, this is really surprising to the degree I could say that maybe this is even a GDB bug, but on the other hand it makes sense to allow for such behavior, especially that the called function may need to interact with other threads itself. So if it wouldn't be a default, this could cause annoying deadlocks.

I think we need to document this behavior/pattern to make sure any future readers of this code understands exactly why we do lock the scheduler.

Alternatively, we can move the `gdb.parse_and_eval` to a `pwndbg.gdblib.*` function, that would be used here instead, and which will use the `lock_scheduler` and we could name and document that function in a descriptive way. We could also put that info into DEVELOPERS.md, as this is really important.

disconnect3d replied:
What @gsingh93 proposed, where we call `set scheduler-locking on` in `ArgparsedCommand` before executing a command also makes sense. Though, it would be great if we can optimize it and only call it if needed.

(We eventually want the commands to be working faster if possible, since they may be used for scripting GDB)

lebr0nli replied:
> Do I understand this correctly that if you call a function like this on behalf of the target in GDB, then if you have other threads those other threads are also resumed for the time this function is executed?

Yes, you are correct.

> Alternatively, we can move the gdb.parse_and_eval to a pwndbg.gdblib.* function

This is a good idea.
What function name would you prefer for this function, and which `pwndbg.gdblib.*` module should I add it to?
cc: @gsingh93 
(I'm not good at naming functions and organizing them into different modules xD)

>  We could also put that info into DEVELOPERS.md, as this is really important.

Sure, I'll add that info into `DEVELOPERS.md`.

gsingh93 replied:
> What function name would you prefer for this function, and which pwndbg.gdblib.* module should I add it to?

I don't think we have an existing module that's a good place for this. And I'm honestly not sure what the module should be called, sorry 😛 

Maybe `parse_and_eval_with_scheduler_lock` for the function name? The verbosity is intentional, it should be clear and stick out when we need to use it. But open to other suggestions.

lebr0nli replied:
> Maybe `parse_and_eval_with_scheduler_lock` for the function name? The verbosity is intentional, it should be clear and stick out when we need to use it. But open to other suggestions.

Ok, thanks!
I put it into `pwndbg.gdblib.sheduler` for now.
Let me know if there’s another good place or good name for the module :p

disconnect3d commented on pwndbg/gdblib/tls.py:
```diff
@@ -14,74 +15,53 @@
+@contextmanager
+def lock_scheduler():
+    already_lock = gdb.parameter("scheduler-locking") == "on"
+    old_config = gdb.parameter("scheduler-locking")
```
Can we fetch the parameter value once?

lebr0nli replied:
Sure, I'll fix it.

CptGibbon commented on pwndbg/heap/ptmalloc.py:
```diff
@@ -1608,218 +1597,189 @@ def thread_cache(self):
+        # We try to find the address by using mp_.srck_base + 0x10 first since it's more reliable than other methods.
+        if not self.multithreaded:
+            try:
+                thread_cache_struct_addr = self.mp["sbrk_base"] + 0x10
```
Just a minor point, `thread_cache_struct_addr` will be inaccurate if the `GLIBC_TUNABLES` environment variable is set.
Here's an example of how pwndbg can deal with this: https://github.com/pwndbg/pwndbg/blob/52a479211cf58d1bdc5b42fd50471f565e6a063c/pwndbg/heap/ptmalloc.py#L1231-L1237
In fact the `get_sbrk_heap_region()` method could be useful here 🤷‍♂️

lebr0nli replied:
Thanks pointing it out, I'll use `get_sbrk_heap_region()` instead.
Btw [here](https://github.com/pwndbg/pwndbg/blob/d4562ea462b8013253c219ed5ddd292386be5c6c/pwndbg/heap/ptmalloc.py#L1139) also used `sbrk_base + 0x10`, should I replace it as well?

CptGibbon replied:
Well spotted, yes that also seems like a good candidate for replacement. It might need some extra modifications since a `Page` object is returned.

disconnect3d replied:
If there are any more improvements to be made here, please lets do it in another PR

lebr0nli commented:
>  To prevent this from happening in the future, I'm wondering if we should always be calling set scheduler-locking on in ArgparsedCommand before a command is executed and set scheduler-locking off when the command is finished. Do you think doing this would cause any problems?

I think this is a good idea, but for the performance, maybe we could only call it only when we need it.
So I'll prefer this solution(https://github.com/pwndbg/pwndbg/pull/1536#discussion_r1083233110) more.

lebr0nli committed:
Refacotr scheduler-locking related functions

- Move these functions into `pwndbg.gdblib.scheduler`
- Fetch the parameter value once (https://github.com/pwndbg/pwndbg/pull/1536#discussion_r1082549746)

lebr0nli committed:
Avoid bug caused by GLIBC_TUNABLES

See https://github.com/pwndbg/pwndbg/pull/1536#discussion_r1083202815

lebr0nli committed:
Add note about `set scheduler-locking on`

disconnect3d commented on pwndbg/gdblib/scheduler.py:
```diff
@@ -0,0 +1,19 @@
+
+@contextmanager
+def lock_scheduler():
+    old_config = gdb.parameter("scheduler-locking")
```
```suggestion
    """
    This context manager can be used to run GDB commands with threads scheduling
    being locked which means that other threads will be stopped during execution.

    This is useful to prevent bugs where e.g.: gdb.parse_and_eval("(int)foo()")
    would execute foo() on the current debugee thread but would also unlock other
    threads for being executed and those other threads may for example hit a
    breakpoint we set previously which would be confusing for the user.

    See also: https://sourceware.org/gdb/onlinedocs/gdb/All_002dStop-Mode.html
    """
    old_config = gdb.parameter("scheduler-locking")
```

lebr0nli committed:
Add comment for `lock_scheduler`

Co-authored-by: Disconnect3d <dominik.b.czarnota@gmail.com>

disconnect3d commented on DEVELOPING.md:
```diff
@@ -108,3 +108,4 @@ Feel free to update the list below!
 
 * We would like to add proper tests for pwndbg - see tests framework PR if you want to help on that.
 
+* If you want to use `gdb.parse_and_eval("a_function_name()")` or something similar that call a function, please remember this might cause another thread to continue execution without `set scheduler-locking on`. If you didn't expect that, you should use `parse_and_eval_with_scheduler_lock` in `pwndbg.gdblib.scheduler` instead.
```
```suggestion
* If you want to use `gdb.parse_and_eval("a_function_name()")` or something similar that call a function, please remember this might cause another thread to continue execution without `set scheduler-locking on`. If you didn't expect that, you should use `parse_and_eval_with_scheduler_lock` from `pwndbg.gdblib.scheduler` instead.
```

disconnect3d committed:
Update DEVELOPING.md
